sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,290 @@
+   * Register the outbound data logical endpoint for the transform alongside it's output value
+   * coder. All registered endpoints will eventually be removed when {@link
+   * #sendBufferedDataAndFinishOutboundStreams()} is called at the end of a bundle.
+   */
+  public <T> void registerOutputDataLocation(String pTransformId, Coder<T> coder) {
+    registerOutputLocation(
+        LogicalEndpoint.data(processBundleRequestIdSupplier.get(), pTransformId), coder);
+  }
+
+  public <T> void registerOutputTimersLocation(
+      String pTransformId, String timerFamilyId, Coder<T> coder) {
+    registerOutputLocation(
+        LogicalEndpoint.timer(processBundleRequestIdSupplier.get(), pTransformId, timerFamilyId),
+        coder);
+  }
+
+  public <T> void registerOutputLocation(LogicalEndpoint endpoint, Coder<T> coder) {
+    LOG.debug("Registering endpoint: {}", endpoint);
+    outputLocations.put(endpoint, coder);
+  }
+
+  public synchronized void flush() throws IOException {
+    Elements.Builder elements = convertBufferForTransmission();
+    if (elements.getDataCount() > 0 || elements.getTimersCount() > 0) {
+      outboundObserver.onNext(elements.build());
+    }
+  }
+
+  /**
+   * Closes the streams for all registered outbound endpoints. Should be called at the end of each
+   * bundle.
+   */
+  public void sendBufferedDataAndFinishOutboundStreams() {
+    LOG.debug("Closing streams for outbound endpoints {}", outputLocations);
+    LOG.debug("Sent outbound data size : {}", perEndpointByteCount);
+    LOG.debug("Sent outbound element count : {}", perEndpointElementCount);
+    if (byteCounter.get() == 0 && outputLocations.isEmpty()) {
+      return;
+    }
+    Elements.Builder bufferedElements = convertBufferForTransmission();
+    for (LogicalEndpoint outputLocation : outputLocations.keySet()) {
+      if (outputLocation.isTimer()) {
+        bufferedElements
+            .addTimersBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setTimerFamilyId(outputLocation.getTimerFamilyId())
+            .setIsLast(true);
+      } else {
+        bufferedElements
+            .addDataBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setIsLast(true);
+      }
+    }
+    outboundObserver.onNext(bufferedElements.build());
+    outputLocations.clear();
+    perEndpointByteCount.clear();
+    perEndpointElementCount.clear();
+  }
+
+  public <T> void acceptData(String pTransformId, T data) throws Exception {
+    accept(LogicalEndpoint.data(processBundleRequestIdSupplier.get(), pTransformId), data);
+  }
+
+  public <T> void acceptTimers(String pTransformId, String timerFamilyId, T timers)
+      throws Exception {
+    accept(
+        LogicalEndpoint.timer(processBundleRequestIdSupplier.get(), pTransformId, timerFamilyId),
+        timers);
+  }
```

Comment:
This will create a new `LogicalEndpoint` object on each timer/element that is output.

Note that the process bundle request id is only needed when we are converting the buffer for transmission, this allows us to return a `FnDataReceiver` from `registerOutputLocation` that writes into a map keyed by `ptransformId`. This is convenient because we can avoid the map lookup and the `LogicalEndpoint` object creation on each element that is output since the returned `FnDataReceiver` can store a reference to the ByteString.Output locally. Finally, this can solve our concurrency issue as well since we can return two variants of this FnDataReceiver, one that synchronizes on a lock object and one that doesn't allowing for the flush method to conditionally lock as well if there is a flushing thread.

We can have a similar map for timers keyed by `ptransformId` and `timerFamilyId` (e.g. `Map<KV<String, String>, ByteString.Output>`).